### ISSUE TITLE
Injected Metric fields for REST API handler significantly degrade the Helidon server's performance

### ISSUE THREAD

**Issue by Pingao2022** [helidon-io/helidon#6825]
[//]: # "Please remove these comments"
[//]: # "Provide a general summary of the issue in the Title above"
I am using the openapi generator tool to auto-generate the REST API handler from a specified OpenAPI Specification (OAS) yaml file. For each of the URLs/paths in the yaml file, the generator tool would generate a unique method to handle the URL. In order to track each URL/path's status code metric information, I also let the generator tool auto-generator some injected metrics for the URL/path. So, in the end, for each URL/path, the auto-generated REST API class would contain something like these:

```   
    @Inject @Metric(name=COUNTER_NAME, description=COUNTER_DESC, absolute=true, tags={"method=GET", "path=/acl-show", "tags=aclShow282", "resp=200"})
    Counter aclShow282200;
    @Inject @Metric(name=COUNTER_NAME, description=COUNTER_DESC, absolute=true, tags={"method=GET", "path=/acl-show", "tags=aclShow282", "resp=400"})
    Counter aclShow282400;
    @Inject @Metric(name=COUNTER_NAME, description=COUNTER_DESC, absolute=true, tags={"method=GET", "path=/acl-show", "tags=aclShow282", "resp=403"})
    Counter aclShow282403;
    @Inject @Metric(name=COUNTER_NAME, description=COUNTER_DESC, absolute=true, tags={"method=GET", "path=/acl-show", "tags=aclShow282", "resp=404"})
    Counter aclShow282404;
    @Inject @Metric(name=COUNTER_NAME, description=COUNTER_DESC, absolute=true, tags={"method=GET", "path=/acl-show", "tags=aclShow282", "resp=500"})
    Counter aclShow282500;
    @Inject @Metric(name=COUNTER_NAME, description=COUNTER_DESC, absolute=true, tags={"method=GET", "path=/acl-show", "tags=aclShow282", "resp=other"})
    Counter aclShow282Other;
    @Inject @Metric(name=HISTOGRAM_NAME, description=HISTOGRAM_DESC, absolute=true, tags={"method=GET", "path=/acl-show", "tags=aclShow282"})
    Histogram aclShow282Hist;

    @GET
    @Path("/acl-show")
    @Produces({ "text/plain" })
    @ApiOperation(value = "get results for acli command (acl-show)", notes = "", response = AcliResults.class, tags={ "CP" })
    @ApiResponses(value = {
        @ApiResponse(code = 200, message = "response containing a list of acli command results from all targeted services/instance", response = AcliResults.class),
        @ApiResponse(code = 400, message = "text description for a bad acli command request response", response = String.class),
        @ApiResponse(code = 500, message = "text description for an internal server error response", response = String.class)
    })
    public Response aclShow282(@QueryParam("service")  @ApiParam("run the acli command on this service; if not present, run the acli command on all")  Service service,@QueryParam("instance")  @ApiParam("run the acli command on this instance")  String instance) {
        Instant start = Instant.now();
        Response response;
        try {
            // get the command from the path
            String fullPath = "/acl-show";
            String[] fullPathElements = fullPath.split("/");

            // join the elements to form the command, replace each path parameter with %s
            StringJoiner cmdFormat = new StringJoiner(" ");
            for (String cmd: fullPathElements) {
                if (cmd.contains("{")) {
                    // a path parameter
                    cmdFormat.add("%s");
                }
                else {
                    // a regular command
                    cmdFormat.add(cmd);
                }
            }

            // get the full command by formatting the command elements with all the path parameters
            String command = String.format(cmdFormat.toString());


            // call the hook
            response = clientMgr.getAcliResults(command,service,instance);
        } catch (Exception e) {
            LOGGER.error(e.getMessage(), e);
            String error = "Internal Server Error";
            response = Response.status(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()).entity(error).build();
        }
        var deltaTime = Duration.between(start, Instant.now()).toMillis();
        aclShow282Hist.update(deltaTime);
        int respStatus = response.getStatus();
        // for stats, keep ALL successful responses in one bucket
        if (respStatus >= Status.OK.getStatusCode()
            && respStatus <= Status.PARTIAL_CONTENT.getStatusCode()) {
            aclShow282200.inc();
        } else if (respStatus == Status.BAD_REQUEST.getStatusCode()) {
            aclShow282400.inc();
        } else if (respStatus == Status.FORBIDDEN.getStatusCode()) {
            aclShow282403.inc();
        } else if (respStatus == Status.NOT_FOUND.getStatusCode()) {
            aclShow282404.inc();
        } else if (respStatus == Status.INTERNAL_SERVER_ERROR.getStatusCode()) {
            aclShow282500.inc();
        } else {
            aclShow282Other.inc();
        }
        return response;
    }
```

The problem:
When the number of URLs/paths increases to more than thousands, the performance of the Helidon server degrades significantly. That is, whenever a HTTP request for a particular URL/path is received, the Helidon server would get stuck, it would take 5 to 10 minutes for the server to eventually pick up the proper API handler method to process the request.

The culprit:
Those injected metric fields for each URL/path seem to be the root cause of the degradation. Once I comment out those metric fields, the performance is again on par, even with thousands of URLs/paths.

I will attach two debugging files:

1. Thread.txt from "jcmd <my Helidon Process id> Thread.print Thread.txt
2. JFRdump.txt from "jcmd <my Helidon Process id> JFR.start duration=100s settings=profile filename=JFRdump.txt

    

## Environment Details
* Helidon Version: Helidon-MP: 3.0.0
* Helidon SE or Helidon MP: MP
* JDK version: oracle-jdk-17:17.0.3.1
* OS: Linux
* Docker version (if applicable): 19.03.11-ol, build a9b9315

----------

## Problem Description
[//]: # "Describe the bug in detail highlighting current behavior vs expected behavior"
Please refer to the summary above

[//]: # "State if the problem is easily reproducible or happens intermittently"
It is easily reproducible

[//]: # "Include stack traces or command outputs"
Please refer to the attached files

## Steps to reproduce
[//]: # "Step by step instructions to reproduce the problem"
Create around one thousand URLs/paths, add some metric fields for each URL, start the server, send a HTTP request to any URL/path, and observe the problem

[//]: # "Provide sample code/application if relevant"
Please refer to the summary above
[Thread.txt](https://github.com/helidon-io/helidon/files/11480009/Thread.txt)
[JFRdump.txt](https://github.com/helidon-io/helidon/files/11480020/JFRdump.txt)

**Comment by tjquinno** [comment]
@Pingao2022 You did not include the entire resource class, but I assume it is either unspecified or `@RequestScoped`, either of which means Helidon creates a new instance for each request. That means each arriving request triggers the creation of another instance of the resource class, which in turn triggers the injection of the metrics you want into the new instance.

There might well be some inefficiencies in how Helidon manages these injections, but here is a possible workaround you can try yourself that might help:

Create an _app-scoped_ bean corresponding to each of the resource classes and `@Inject` the relevant metrics on _that_ app-scoped bean. Then `@Inject` the appropriate app-scoped bean into the resource class. That will be only one `@Inject` per request instead of one for each of the metrics.

**Comment by tjquinno** [comment]
I also see that Santiago has assigned this to himself earlier today. He might have some other suggestions or insights.

**Comment by spericas** [comment]
@Pingao2022 Would it be possible for you to share either a complete reproducible, or just the generated classes, or at least a single class that we can be used as a template to create 1000+ of them?

**Comment by spericas** [comment]
@Pingao2022 I found a problem in our code that is likely related to the performance issue that you uncovered. Would it be possible for you try a SNAPSHOT version of Helidon? PR #6842 contains the fix. You'd need to build from source (using the `helidon-3.x` branch) and point your app to the SNAPSHOT version.

**Comment by Pingao2022** [comment]
> @Pingao2022 You did not include the entire resource class, but I assume it is either unspecified or `@RequestScoped`, either of which means Helidon creates a new instance for each request. That means each arriving request triggers the creation of another instance of the resource class, which in turn triggers the injection of the metrics you want into the new instance.
> 
> There might well be some inefficiencies in how Helidon manages these injections, but here is a possible workaround you can try yourself that might help:
> 
> Create an _app-scoped_ bean corresponding to each of the resource classes and `@Inject` the relevant metrics on _that_ app-scoped bean. Then `@Inject` the appropriate app-scoped bean into the resource class. That will be only one `@Inject` per request instead of one for each of the metrics.

@tjquinno 
You were right. No scope was provided for the over-all REST API class.
I added an ApplicationScoped annotation to the class, the performance seemed to be improved, but it still took relatively long time to get the response back
Will keep the "ApplicationScoped" annotation from now on

**Comment by Pingao2022** [comment]
> @Pingao2022 I found a problem in our code that is likely related to the performance issue that you uncovered. Would it be possible for you try a SNAPSHOT version of Helidon? PR #6842 contains the fix. You'd need to build from source (using the `helidon-3.x` branch) and point your app to the SNAPSHOT version.

@spericas 
Thanks. Will try it out and update later

**Comment by Pingao2022** [comment]
Fix (PR #6842) is working.
The response time now is normal even with thousands of URLs, each with injected metrics.
We need this fix ASAP, so, please make it to the official release ASAP.

Thanks for all the helps!
